src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -398,18 +387,14 @@ private int checkArraySize(String typeName, int elemSize) {
         return (int)arraySize;
     }
 
-    private void checkBounds(long offset, long length) {
-        if (isSmall() &&
-                offset <= Integer.MAX_VALUE && length <= Integer.MAX_VALUE &&
-                offset >= Integer.MIN_VALUE && length >= Integer.MIN_VALUE) {
-            checkBoundsSmall((int)offset, (int)length);
-        } else if (this != NativeMemorySegmentImpl.EVERYTHING) { // oob not possible for everything segment
-            if (
-                    length < 0 ||
-                    offset < 0 ||
-                    offset > this.length - length) { // careful of overflow
-                throw outOfBoundException(offset, length);
-            }
+    @ForceInline
+    void checkBounds(long offset, long length) {
+        if (length > 0) {
```

Comment:
This is the best I could do, given that `Objects.checkIndex` implements a <= semantics on the upper bound. That is, we need to protect the code from overflow, which can happen if `length == 0` (which can happen in slicing). When that happens, the upper bound in the second check `this.length - length + 1` ends up being `this.length + 1`, which is problematic. I've also tested to see if we could just use the old check bounds routine, but it seems like using `Objects.checkIndex` is required to properly apply BCE.

Reply:
I think `Objects.checkIndex` checks if `length` && `offset` are negative, so there's no risk of overflow (actually it checks if `length` is negative, and then make unsigned compare), so maybe I miss here something.

Just minor, I wonder if it will be better to use `Preconditions.checkIndex` - one method less to inline.

Reply:
Consider a case of a segment whose length is Long.MAX_VALUE. Then you want to take a slice of size 0. This will throw an exception, as Long.MAX_VALUE + 1 will flip into negative territory.

Another problem is when the segment has size zero, in which case the check:

```
long checkedOffset = Objects.checkIndex(offset, this.length);
```

Will also fail. To make that pass we'd need to tweak that check to use `this.length + 1` again, and risk overflow. So I figured it was better just just special case when the accessed length was zero (which should happen rarely). Another option would be to have different checks for slicing and for access (since for access length can never be zero) - but then it duplicates the implementation, and there are also methods like `copy` which are left a bit in the middle.

As for inlining, I doubt that it would make a difference to call Preconditions directly, since it's all `@ForceInlined`.

Reply:
I get it, it's to allow creating 0-length chunks (I've thought in context of reading value).

If we assume `length > 0`, then `this.length - length + 1 =< this.length`, (in case `this.len == 0 -> 0 <= 0`)

Reply:
I'm not sure if I missed some case, but if it's in if (`length > 0`), then one `checkIndex` is redundant? Unless we want to disallow slicing 0 length indices?

Reply:
@rsmogura you are correct, the extra check is redundant now - we only need the second. I probably added before I had the check for `length > 0` and then left that there. Effectively, the second check is enough, because that checks that `offset >= 0`,  and `length > 0` by construction (and we now that segment size must be `>= 0`). So I believe we should be ok.

@merykitty yes, this is implementing the same check in spirit as `checkFromIndexSize` - but unfortunately that is not a VM intrinsics, so we have to code around that.